workspaces/ui-v2/src/optic-components/pages/diffs/AddEndpointsPage/components/BulkLearnModal.tsx

```diff
@@ -0,0 +1,150 @@
+import { IUndocumentedUrl } from '<src>/optic-components/hooks/diffs/SharedDiffState';
+import { makePattern } from '../utils';
+
+type BulkLearnModalProps = {
+  undocumentedEndpointsToLearn: IUndocumentedUrl[];
+  closeModal: () => void;
+};
+
+const LearningProgress: FC<{
+  pendingEndpointIds: Set<string>;
+  closeModal: BulkLearnModalProps['closeModal'];
+}> = ({ pendingEndpointIds, closeModal }) => {
+  const { stageEndpoint, context } = useSharedDiffContext();
+  const refStageEndpoint = useRef(stageEndpoint);
+
+  // @nic TODO handle when learning fails! - currently the modal just gets stuck :(
+  const learningPendingEndpoints = context.pendingEndpoints.filter(
+    (pendingEndpoint) => pendingEndpointIds.has(pendingEndpoint.id)
+  );
+  const completedCount = learningPendingEndpoints.filter((pendingEndpoint) =>
+    pendingEndpoint.ref.state.matches('ready')
+  ).length;
+  const totalCount = learningPendingEndpoints.length;
+
+  // Stage endpoints and close modal when we've completed all
+  useEffect(() => {
+    if (completedCount === totalCount) {
+      for (const pendingId of pendingEndpointIds) {
+        refStageEndpoint.current(pendingId);
+      }
+      closeModal();
+    }
+  }, [completedCount, totalCount, pendingEndpointIds, closeModal]);
+
+  // This is a lie since we have no guarantee these are learnt in order
+  // But it's nice to give visual feedback to users and see some progress happening
+  const currentlyProcessingEndpoint =
+    learningPendingEndpoints[Math.min(completedCount, totalCount - 1)];
+
+  return (
+    <Box display="flex">
+      <Box display="flex" padding="0 8px" alignItems="center">
+        <SpinningOpticLogo />
+      </Box>
+      <Box width="100%" paddingRight="16px">
+        Learning endpoint ({completedCount}/{totalCount}):{' '}
+        <EndpointName
+          method={currentlyProcessingEndpoint.method}
+          leftPad={0}
+          fontSize={12}
+          fullPath={currentlyProcessingEndpoint.pathPattern}
+        />
+        <LinearProgress variant="indeterminate" />
+      </Box>
+    </Box>
+  );
+};
+
+export const BulkLearnModal: FC<BulkLearnModalProps> = ({
+  undocumentedEndpointsToLearn,
+  closeModal,
+}) => {
+  const { documentEndpoint, wipPatterns } = useSharedDiffContext();
+  const endpointsAsPatterns = undocumentedEndpointsToLearn.map(
+    ({ path, method }) => {
+      // For selectedUrls, unspecified selected urls will not be
+      // in wipPatterns
+      const pattern = wipPatterns[path + method]
+        ? makePattern(wipPatterns[path + method].components)
+        : path;
+      return { pattern, method };
+    }
+  );
+  const [learningInfo, setLearningInfo] = useState<{
+    pendingEndpointIds: Set<string>;
+  } | null>(null);
+
+  const learnEndpoints = () => {
+    const pendingEndpointIds = endpointsAsPatterns.map(({ pattern, method }) =>
+      documentEndpoint(pattern, method)
```

Comment:
This dumps all the pending endpoints onto the learn worker - in the local-cli instance this will be slow, but we've already implemented a separate PR to bottleneck this https://github.com/opticdev/optic/pull/750